README.md
yii2-oauth2-server
==================

A wrapper for implementing an
[OAuth2 Server](https://github.com/bshaffer/oauth2-server-php).

[![Latest Stable Version](https://poser.pugx.org/tecnocen/yii2-oauth2-server/v/stable)](https://packagist.org/packages/tecnocen/yii2-oauth2-server)
[![Total Downloads](https://poser.pugx.org/tecnocen/yii2-oauth2-server/downloads)](https://packagist.org/packages/tecnocen/yii2-oauth2-server)
[![Code Coverage](https://scrutinizer-ci.com/g/tecnocen-com/yii2-oauth2-server/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/tecnocen-com/yii2-oauth2-server/?branch=master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/tecnocen-com/yii2-oauth2-server/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/tecnocen-com/yii2-oauth2-server/?branch=master)

Scrutinizer [![Build Status Scrutinizer](https://scrutinizer-ci.com/g/tecnocen-com/yii2-oauth2-server/badges/build.png?b=master&style=flat)](https://scrutinizer-ci.com/g/tecnocen-com/yii2-oauth2-server/build-status/master)
Travis [![Build Status Travis](https://travis-ci.org/tecnocen-com/yii2-oauth2-server.svg?branch=master&style=flat?style=for-the-badge)](https://travis-ci.org/tecnocen-com/yii2-oauth2-server)

This project was forked from
[Filsh Original Project](https://github.com/Filsh/yii2-oauth2-server) but the
changes are not transparent, read [UPGRADE.md] to pass to the latest version.

Installation
------------

The preferred way to install this extension is through [composer](http://getcomposer.org/download/).

Either run

```
php composer.phar require --prefer-dist tecnocen/yii2-oauth2-server "*"
```

or add

```json
"tecnocen/yii2-oauth2-server": "~2.1"
```

to the require section of your composer.json.

Usage
-----

To use this extension,  simply add the following code in your application configuration as a new module:

```php
    'bootstrap' => ['oauth2'],
    'modules'=>[
        //other modules .....
        'oauth2' => [
            'class' => 'tecnocen\oauth2server\Module',            
            'tokenParamName' => 'accessToken',
            'tokenAccessLifetime' => 3600 * 24,
            'storageMap' => [
                'user_credentials' => 'app\models\User',
            ],
            'grantTypes' => [
                'user_credentials' => [
                    'class' => 'OAuth2\GrantType\UserCredentials',
                ],
                'refresh_token' => [
                    'class' => 'OAuth2\GrantType\RefreshToken',
                    'always_issue_new_refresh_token' => true
                ]
            ]
        ]
    ],
```

Bootstrap will initialize translation and add the required url rules to
`Yii::$app->urlManager`.

### JWT tokens

There is no JWT token support on this fork, feel free to submit a
(pull request)[https://github.com/tecnocen-com/yii2-oauth2-server/pulls] to
enable this functionality.

### UserCredentialsInterface

The class passed to `Yii::$app->user->identityClass` must implement the interface
`\OAuth2\Storage\UserCredentialsInterface`, to store oauth2 credentials in user
table.

```php
use Yii;

class User extends common\models\User
    implements \OAuth2\Storage\UserCredentialsInterface
{

    /**
     * Implemented for Oauth2 Interface
     */
    public static function findIdentityByAccessToken($token, $type = null)
    {
        /** @var \tecnocen\oauth2server\Module $module */
        $module = Yii::$app->getModule('oauth2');
        $token = $module->getServer()->getResourceController()->getToken();
        return !empty($token['user_id'])
                    ? static::findIdentity($token['user_id'])
                    : null;
    }

    /**
     * Implemented for Oauth2 Interface
     */
    public function checkUserCredentials($username, $password)
    {
        $user = static::findByUsername($username);
        if (empty($user)) {
            return false;
        }
        return $user->validatePassword($password);
    }

    /**
     * Implemented for Oauth2 Interface
     */
    public function getUserDetails($username)
    {
        $user = static::findByUsername($username);
        return ['user_id' => $user->getId()];
    }
}
```

### Migrations

The next step your shold run migration

```php
yii migrate all -p=@tecnocen/oauth2server/migrations/tables
yii fixture "*" -n=tecnocen/oauth2server/fixtures
```

this migration create the oauth2 database scheme. The second command insert
test user credentials ```testclient:testpass``` for ```http://fake/```

### Controllers

To support authentication by access token. Simply add the behaviors for your controller or module.

```php
use yii\helpers\ArrayHelper;
use yii\filters\auth\HttpBearerAuth;
use yii\filters\auth\QueryParamAuth;
use tecnocen\oauth2server\filters\auth\CompositeAuth;

class Controller extends \yii\rest\Controller
{
    /**
     * @inheritdoc
     */
    public function behaviors()
    {
        return ArrayHelper::merge(parent::behaviors(), [
            'authenticator' => [
                'class' => CompositeAuth::class,
                'authMethods' => [
                    ['class' => HttpBearerAuth::class],
                    [
                        'class' => QueryParamAuth::class,
                        'tokenParam' => 'accessToken',
                    ],
                ]
            ],
        ]);
    }
}
```

The code above is the same as the default implementation which can be
simplified as

```php
use yii\helpers\ArrayHelper;
use tecnocen\oauth2server\filters\auth\CompositeAuth;

class Controller extends \yii\rest\Controller
{
    /**
     * @inheritdoc
     */
    public function behaviors()
    {
        return ArrayHelper::merge(parent::behaviors(), [
            'authenticator' => CompositeAuth::class,
        ]);
    }
}
```

### Scopes

The property `tecnocen\oauth2server\filters\auth\CompositeAuth::$actionScopes`
set which actions require specific scopes. If those scopes are not meet the
action wont be executed, and the server will reply with an HTTP Status Code 403.

```php
public function behaviors()
{
    return ArrayHelper::merge(parent::behaviors(), [
        'authenticator' => [
            'class' => CompositeAuth::class,
            'actionScopes' => [
                'create' => 'default create',
                'update' => 'default edit',
                '*' => 'default', // wildcards are allowed
            ]
        ],,
    ]);
}
```

### Automatically Revoke Tokens

Sometimes its neccessary to revoke a token on each request to prevent the
request from being triggered twice.

To enable this functionality you need to implement
`tecnocen\oauth2server\RevokeAccessTokenInterface` in the class used to identify
the authenticated user.

```php
use OAuth2\Storage\UserCredentialsInterface;
use tecnocen\oauth2server\RevokeAccessTokenInterface;
use tecnocen\oauth2server\RevokeAccessTokenTrait;

class User extend \yii\db\ActiveRecord implement
    UserCredentialsInterface,
    RevokeAccessTokenInterface
{
    use tecnocen\oauth2server\RevokeAccessTokenTrait; // optional, trait with default implementation.
    
    // rest of the class.
}
```
Then use the previous class as configuration for `Yii::$app->user->identityClass`

Attaching the action filter `tecnocen\oauth2server\filters\RevokeAccessToken`
allows to configure the actions to automatically revoke the access token.

```php
public function behaviors()
{
    return [
        'revokeToken' => [
            'class' => \tecnocen\oauth2server\filters\RevokeAccessToken::class,
            // optional only revoke the token if it has any of the following
            // scopes. if not defined it will always revoke the token.
            'scopes' => ['author', 'seller'],
            // optional whether or not revoke all tokens or just the active one
            'revokeAll' => true,
            // optional if non authenticated users are permited.
            'allowGuests' => true,
            // which actions this behavior applies to.
            'only' => ['create', 'update'],
        ]
    ];
}
```

### Generate Token with JS

To get access token (js example):

```js
var url = window.location.host + "/oauth2/token";
var data = {
    'grant_type':'password',
    'username':'<some login from your user table>',
    'password':'<real pass>',
    'client_id':'testclient',
    'client_secret':'testpass'
};
//ajax POST `data` to `url` here
//
```

## Built With

* Yii 2: The Fast, Secure and Professional PHP Framework [http://www.yiiframework.com](http://www.yiiframework.com)

## Code of Conduct

Please read [CODE_OF_CONDUCT.md](https://github.com/tecnocen-com/yii2-oauth2-server/blob/master/CODE_OF_CONDUCT.md) for details on our code of conduct.

## Contributing

Please read [CONTRIBUTING.md](https://github.com/tecnocen-com/yii2-oauth2-server/blob/master/CONTRIBUTING.md) for details on the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/tecnocen-com/yii2-oauth2-server/tags).

_Considering [SemVer](http://semver.org/) for versioning rules 9, 10 and 11 talk about pre-releases, they will not be used within the Tecnocen-com._

## Authors

* [**Angel Guevara**](https://github.com/Faryshta) - *Initial work* - [Tecnocen.com](https://github.com/Tecnocen-com)
* [**Carlos Llamosas**](https://github.com/neverabe) - *Initial work* - [Tecnocen.com](https://github.com/Tecnocen-com)

See also the list of [contributors](https://github.com/tecnocen-com/yii2-oauth2-server/graphs/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* TO DO - Hat tip to anyone who's code was used
* TO DO - Inspiration
* TO DO - etc

[![yii2-oauth2-server](https://img.shields.io/badge/Powered__by-Tecnocen.com-orange.svg?style=for-the-badge)](https://www.tecnocen.com/)

For more, see https://github.com/bshaffer/oauth2-server-php


CHANGELOG.md
Yii2 OAuth2 Server Library
==========================

4.1.1
-----

- [Bug] Fixed support for I18N and added 'es-MX' translation.
  (Faryshta)

4.1.0
-----

- [Enh] Functionality to revoke a token after it gets used by
  `tecnocen\oauth2server\filters\RevokeAccessToken` and
  `tecnocen\oauth2server\RevokeAccessTokenInterface`

4.0.1
-----

- [Bug] Add action for OPTIONS verb on
  `tecnocen\oauth2server\controllers\RestController` (Faryshta)


4.0.0
-----

- [Brk] `tecnocen\oauth2Server\filters\auth\CompositeAuth` checks for errors on
  the oauth2 request simplifying the use to just one behavior
- [Enh] Live Demo on `tests/_app` initialized by command
  `tests/_app/yii.php serve`
- [Brk] Support for `OPTIONS` requests needed on CORS.
- [Enh] `tecnocen\oauth2Server\filters\auth\CompositeAuth::$actionScopes`
  allows to set which scopes a token must have for determined actions.


3.0.2
-----

- [Bug] `tecnocen\oauth2Server\controllers\RestController::behaviors()`
  obtains the `oauth2Module` from parent module.


3.0.1
-----

- [Bug] Initialize oauth2Module on `EVENT_BEFORE_ACTION`.

3.0.0
-----

- [BRK] Taking over the project changing the namespaces.
- [Enh] Create codeception api tests.
- [Enh] Initialize oauth2 translations using `yii\base\Application::$bootstrap`
- [Enh] Initialize oauth2 server using `yii\base\Module::EVENT_BEFORE_ACTION`
- [Enh] Update migrations
- [Enh] Support for yii2 version 2.0.13


src/Module.php
<?php

namespace tecnocen\oauth2server;

use OAuth2\Request;
use OAuth2\Response;
use ReflectionClass;
use Yii;
use yii\base\BootstrapInterface;
use yii\base\InvalidConfigException;
use yii\i18n\PhpMessageSource;
use yii\web\UrlRule;

/**
 * For example,
 *
 * ```php
 * 'oauth2' => [
 *     'class' => 'tecnocen\oauth2server\Module',
 *     'tokenParamName' => 'accessToken',
 *     'tokenAccessLifetime' => 3600 * 24,
 *     'storageMap' => [
 *         'user_credentials' => 'common\models\User',
 *     ],
 *     'grantTypes' => [
 *         'user_credentials' => [
 *             'class' => 'OAuth2\GrantType\UserCredentials',
 *         ],
 *         'refresh_token' => [
 *             'class' => 'OAuth2\GrantType\RefreshToken',
 *             'always_issue_new_refresh_token' => true
 *         ]
 *     ]
 * ]
 * ```
 */
class Module extends \yii\base\Module implements BootstrapInterface
{
    /**
     * @var bool whether the oauth2 server was initialized
     */
    private $serverInitialized = false;

    /**
     * @inheritdoc
     */
    public $controllerNamespace = controllers::class;

    /**
     * @var array Model's map
     */
    public $modelMap = [];

    /**
     * @var array Storage's map
     */
    public $storageMap = [];

    /**
     * @var array GrantTypes collection
     */
    public $grantTypes = [];

    /**
     * @var string name of access token parameter
     */
    public $tokenParamName;

    /**
     * @var type max access lifetime
     */
    public $tokenAccessLifetime;
    /**
     * @var array Model's map
     */
    protected $defaultModelMap = [
        'OauthClients' => models\OauthClients::class,
        'OauthAccessTokens' => models\OauthAccessTokens::class,
        'OauthAuthorizationCodes' => models\OauthAuthorizationCodes::class,
        'OauthRefreshTokens' => models\OauthRefreshTokens::class,
        'OauthScopes' => models\OauthScopes::class,
    ];

    /**
     * @var array Storage's map
     */
    protected $defaultStorageMap = [
        'access_token' => storage\Pdo::class,
        'authorization_code' => storage\Pdo::class,
        'client_credentials' => storage\Pdo::class,
        'client' => storage\Pdo::class,
        'refresh_token' => storage\Pdo::class,
        'user_credentials' => storage\Pdo::class,
        'public_key' => storage\Pdo::class,
        'jwt_bearer' => storage\Pdo::class,
        'scope' => storage\Pdo::class,
    ];

    /**
     * @inheritdoc
     */
    public function urlRules()
    {
        return [
            [
                'class' => UrlRule::class,
                'pattern' => $this->getUniqueId() . '/<action:\w+>',
                'route' => $this->getUniqueId() . '/rest/<action>',
                'verb' => ['POST'],
            ],
            [
                'class' => UrlRule::class,
                'pattern' => $this->getUniqueId() . '/<action:\w+>',
                'route' => $this->getUniqueId() . '/rest/options',
                'verb' => ['OPTIONS'],
            ],
        ];
    }

    /**
     * Initializes the oauth2 server and its dependencies.
     */
    public function initOauth2Server()
    {
        if ($this->serverInitialized) {
            return;
        }

        $this->serverInitialized = true;
        $this->modelMap = array_merge($this->defaultModelMap, $this->modelMap);
        $this->storageMap = array_merge($this->defaultStorageMap, $this->storageMap);
        foreach ($this->modelMap as $name => $definition) {
            Yii::$container->set(models::class . '\\' . $name, $definition);
        }

        foreach ($this->storageMap as $name => $definition) {
            Yii::$container->set($name, $definition);
        }

        $storages = [];
        foreach(array_keys($this->storageMap) as $name) {
            $storages[$name] = Yii::$container->get($name);
        }

        $grantTypes = [];
        foreach($this->grantTypes as $name => $options) {
            if(!isset($storages[$name]) || empty($options['class'])) {
                throw new InvalidConfigException(
                    'Invalid grant types configuration.'
                );
            }

            $class = $options['class'];
            unset($options['class']);

            $reflection = new ReflectionClass($class);
            $config = array_merge([0 => $storages[$name]], [$options]);

            $instance = $reflection->newInstanceArgs($config);
            $grantTypes[$name] = $instance;
        }

        $this->set('server', Yii::$container->get(Server::class, [
            $this,
            $storages,
            [
                'token_param_name' => $this->tokenParamName,
                'access_lifetime' => $this->tokenAccessLifetime,
            ],
            $grantTypes
        ]));
        $this->set('request', Request::createFromGlobals());
        $this->set('response', new Response());
    }

    /**
     * @inheritdoc
     */
    public function beforeAction($action)
    {
        if (!parent::beforeAction($action)) {
            return false;
        }
        $this->initOauth2Server();

        return true;
    }

    /**
     * @inheritdoc
     */
    public function bootstrap($app)
    {
        if ($app instanceof \yii\web\Application) {
            $app->getUrlManager()->addRules($this->urlRules());
        } else {
            $this->controllerNamespace = commands::class;
        }

        $this->registerTranslations($app);
    }

    /**
     * Gets Oauth2 Server
     *
     * @return Server
     */
    public function getServer()
    {
        return $this->get('server');
    }

    /**
     * Gets Oauth2 Response
     *
     * @return Response
     */
    public function getResponse()
    {
        return $this->get('response');
    }

    /**
     * Gets Oauth2 Request
     *
     * @return Request
     */
    public function getRequest()
    {
        return $this->get('request');
    }

    /**
     * Register translations for this module
     */
    public function registerTranslations($app)
    {
        if(!isset($app->get('i18n')->translations['tecnocen/oauth2/*'])) {
            $app->get('i18n')->translations['tecnocen/oauth2/*'] = [
                'class'    => PhpMessageSource::class,
                'basePath' => __DIR__ . '/messages',
                'fileMap' => [
                    'tecnocen/oauth2/oauth2server' => 'oauth2server.php',
                ],
            ];
        }
    }

    /**
     * Translate module message
     *
     * @param string $category
     * @param string $message
     * @param array $params
     * @param string $language
     * @return string
     */
    public static function t(
        $category,
        $message,
        $params = [],
        $language = null
    ) {
        return Yii::t(
            'tecnocen/oauth2/' . $category,
            $message,
            $params,
            $language
        );
    }
}
